### PR TITLE
RD2: Create next Opportunity for RD having Closed Lost Opp in migration

### PR DESCRIPTION
**Pull request by RadSFDeveloper**
Enable automated next Opportunity creation when the same next Close Date Opp is Closed Lost in data migration.

# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by RadSFDeveloper** [comment]
**lurch: attach W-038554

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HqaDQAS" target="_blank">W-038554</a>

**Comment by force2b** [comment]
@RadSFDeveloper I'm a bit confused on this PR. There are 61 commits and 37 changed files. Isn't this the item we just spoke about yesterday afternoon? I was expected maybe 3 or 4 changed files at the most.

**Comment by RadSFDeveloper** [comment]
> @RadSFDeveloper I'm a bit confused on this PR. There are 61 commits and 37 changed files. Isn't this the item we just spoke about yesterday afternoon? I was expected maybe 3 or 4 changed files at the most.

I needed to apply one extra step when PR #4903 got merged, ie merge `feature/RD2` into this branch. It has been done.

File: src/classes/RD2_DataMigrationBase_BATCH.cls [hunk 1]
```diff
@@ -41,6 +41,8 @@ public abstract class RD2_DataMigrationBase_BATCH implements Database.Batchable<
     public static final String LOG_CONTEXT_DRY_RUN = 'RDValidateMigration:';
     public static final String LOG_EXCEPTION_ERROR_TYPE = 'Data Migration Exception';
 
+    public static final Integer MIGRATION_INSTALLMENT_NUMBER = RD2_OpportunityService.MIGRATION_INSTALLMENT_NUMBER;
```
Comment (force2b) [inline src/classes/RD2_DataMigrationBase_BATCH.cls]: Can you add a block comment here explaining what this represents?

File: src/classes/RD2_OpportunityService.cls [hunk 2]
```diff
@@ -36,6 +36,8 @@
 */
 public with sharing class RD2_OpportunityService {
 
+    public static final Integer MIGRATION_INSTALLMENT_NUMBER = 0;
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls]: Can you add a block comment here explaining what this represents?
Also, wondering if a var name change is warranted? Maybe something like `MIGRATION_INSTALLMENT_NUMBER_FLAG` or `MIGRATION_INSTALLMENT_CLOSED_FLAG`?

File: src/classes/RD2_OpportunityService.cls [hunk 3]
```diff
@@ -141,15 +143,19 @@ public with sharing class RD2_OpportunityService {
 
 
     /***
-    * @description Indicates if the Opportunity already exists
+    * @description Indicates if the Opportunity should be created
```
Comment (force2b) [inline src/classes/RD2_OpportunityService.cls]: Can you update the description here to explain why it's checking for a specific value in the `Recurring_Donation_Installment_Number__c` field? If anyone else (or even any of us) look at this code in 6 months we're not going to remember why this is being done and what the business use case was.

File: src/classes/RD2_DataMigrationBase_BATCH.cls [hunk 4]
```diff
@@ -316,6 +318,8 @@ public abstract class RD2_DataMigrationBase_BATCH implements Database.Batchable<
         } else if (rdSettings.npe03__Open_Opportunity_Behavior__c == RD2_Constants.CloseActions.Mark_Opportunities_Closed_Lost.name()) {
             for (Opportunity opp : opps) {
                 opp.StageName = System.Label.npe03.RecurringDonationClosedLostOpportunityStage;
+                opp.Recurring_Donation_Installment_Number__c = MIGRATION_INSTALLMENT_NUMBER;
```
Comment (force2b) [inline src/classes/RD2_DataMigrationBase_BATCH.cls]: Might be worth updating the method comment to explain why we're setting this field value in additional to changing the stage to ClosedLost.

File: src/classes/RD2_OpportunityService.cls (hunk 2 above)
Comment (RadSFDeveloper) [reply src/classes/RD2_OpportunityService.cls]: Code is ready for review. Do we foresee the RD Installment Number being used to exclude (ignore) Opps by different processes not just data migration?

Comment (force2b) [reply src/classes/RD2_OpportunityService.cls]: I don't foresee that.

**Comment by RadSFDeveloper** [comment]
@Julian88Tex , the PR is ready for testing.

**Comment by Julian88Tex** [comment]
Testing has been completed on this WI: https://foundation.lightning.force.com/lightning/r/agf__ADM_Work__c/a2x1E000001HqaDQAS/view

**Pull request merged**